components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java

```diff
@@ -15,323 +15,189 @@
-        this.stats = recordStats ? new ConnectionPoolStats() : NULL_CONNECTION_POOL_STATS;
-    }
-
-    private static <T> void removeEachAndProcess(Queue<T> queue, Consumer<T> consumer) {
-        while (true) {
-            T item = queue.poll();
-
-            if (item == null) {
-                break;
-            }
-
-            consumer.accept(item);
-        }
-    }
-
-    @Override
-    public int compareTo(ConnectionPool other) {
-        return this.origin.compareTo(other.getOrigin());
     }
 
-    @Override
     public Origin getOrigin() {
         return origin;
     }
 
     @Override
     public Observable<Connection> borrowConnection() {
-        if (busyConnections.incrementAndGet() <= connectionPoolSettings.maxConnectionsPerHost()) {
-            return doBorrowConnection();
-        }
-
-        busyConnections.decrementAndGet();
-
-        TimeoutObservableFactory timeoutObservableFactory = new TimeoutObservableFactory(origin, waitingSubscribers);
-        return timeoutObservableFactory.create(stats, connectionPoolSettings);
+        return toObservable(borrowConnection2());
     }
 
-    @Override
-    public boolean returnConnection(Connection connection) {
-        stats.recordTimeToFirstByte(connection.getTimeToFirstByteMillis());
-        if (connection.isConnected()) {
-            Subscriber<? super Connection> subscriber = waitingSubscribers.poll();
-            if (subscriber != null) {
-                subscriber.onNext(connection);
-                subscriber.onCompleted();
+    public Publisher<Connection> borrowConnection2() {
+        return Mono.<Connection>create(sink -> {
+            Connection connection = dequeue();
+            if (connection != null) {
+                if (borrowedCount.getAndIncrement() < poolSettings.maxConnectionsPerHost()) {
+                    sink.success(connection);
+                } else {
+                    borrowedCount.decrementAndGet();
+                    queueNewConnection(connection);
+                }
             } else {
-                // todo this synchronization can probably be removed
-                synchronized (connection) {
-                    busyConnections.decrementAndGet();
-                    borrowedConnections.remove(connection);
-                    availableConnections.add(connection);
+                if (waitingSubscribers.size() >= poolSettings.maxPendingConnectionsPerHost()) {
+                    sink.error(new MaxPendingConnectionsExceededException(
+                            origin,
+                            poolSettings.maxPendingConnectionsPerHost(),
+                            poolSettings.maxPendingConnectionsPerHost()));
+                } else {
+                    this.waitingSubscribers.add(sink.onCancel(() -> waitingSubscribers.remove(sink)));
+                    if (borrowedCount.get() < poolSettings.maxConnectionsPerHost()) {
+                        newConnection();
```

Comment:
If you look carefully, the `newConnection` triggers a connection establishment and publishes a newly created connection either to a waiting subscriber or queues it in `activeConnections` queue.

But I see now we should consider some kind of back-off policy in case connection establishment fails.

Reply:
OK, my confusion was because of this method `private Observable<Connection> newConnection(int attempts)` having a return type but now I see that it calls itself recursively.